### PR TITLE
Cleanup 4way-if

### PR DESCRIPTION
**Pull request by ledvinap**
4way-if is working on silabs BLHeli bootloader and AVR SimonK bootloader

Can someone test AVR-BLHeli bootloader? I'll get to it later ..

### CODE REVIEW

**Comment by ledvinap** [comment]
Works with BLHeli bootloader on Atmel

**Comment by ledvinap** [comment]
Ready for merge.

All comments welcome

**Comment by 4712** [comment]
Thanks for all the work and thanks for finding the stk500v2 StkRcvPacket bug, which lets read of 253 bytes fail. (Btw: the failure never occurs in normal operation).
Is there any additional improvement in the 4-way functionality you found?

**Comment by ledvinap** [comment]
@4712: There was minor buffer overflow in StkRcvPacket (ack bytes are stored after buffer), but in real code it overwrites ESC and possibly I_PARAM_LEN (it looked as if 2 bytes were received after data, but only 1   is documented.  .. I didn't investigate it further, maybe I just overlooked something).
Otherwise the code looks good. I mostly changed coding style and refactored some parts.
I can undo most changes to bootloader drivers if that helps you keeping them in sync with Arduino/AVR code.
I would like to keep esc4wayProcessCmd and esc4wayProcess separated, so that packet framing can be easily replaced.

**Comment by ledvinap** [comment]
@4712 : Can you consider using MSP framing for Cleanflight 4way interface? Probably packing 4way command/address and data as one MSP command.
BLHeliSuite does implement MSP framing, so it should be quite easy.

Only problem is buffer size - MSP input buffer is only 64 bytes now. We can increase buffer size, but limiting MSP packets size will be better in long term (I would prefer to limit MSP packets to about 80B (64B payload + 16B for headers), but even 128B (144B) may be reasonable.

It would allow removing 4way framing, saving some FLASH space and slightly decreasing overall code complexity.

**Comment by 4712** [comment]
MSP framing doesn't make sense to me. Buffer size reduction might be doable. But why, I thought the buffer is only consuming RAM while the interface is active...
Well, I asked questions enough about the flash space consequences of implementing the 4way stuff over and over again... no reactions. Now I did a lot of work and things get stuck because of size matters ... 

On the long run it might be better to remove the whole 4way stuff and get back to a non buffer consuming 1wire pass-through for serial rs232 and the st500v2 protocol as well.
Less space consuming, and more universal usage. There are downsides as well...

**Comment by ledvinap** [comment]
@4712 : Personally I thing that 4way interface is well worth the flash. It's much more elegant and maintainable than 1wire. Pass-through command may be added for universal access to ESC.

My idea was to use existing MSP command handling instead of 4Way framing. With current code there is no need to exit MSP processing (unless I overlooked something), it should be possible to implement 4way without switching to 4way processing loop.
There is persistent 64B input buffer per MSP connection (and unlimited output stream). This size may be easily increased, but limiting MSP command size has long-term advantages (MSP may be implemented over air/i2c  / MSP multiplexing with blackbox etc)

**Comment by hydra** [comment]
@4712 I thought I saw and answered your questions regarding flash place and we decided it was a very useful feature and that we can do optimizations at a later date to make everything fit.  @ledvinap is talking now about such optimizations.  Perhaps I missed something though.

Any and all code in the codebase can be optimized as we see fit,  it makes sense to address @ledvinap 's points now because there is other on-going cleanup work going on with the MSP code.

Anything that you can do to help is greatly appreciated!

**Comment by 4712** [comment]
Yes for Cleanup, but no for restructuring of the protocol. I will not add MSP framing into BLHeliSuite, not ATM. Too  many other things in progress...
BTW, I'm not sure about side effects, when MSP is still enabled while flashing an ESC.

**Comment by ledvinap** [comment]
@4712: I thought that MSP is already implemented in BLHeliSuite (to get cleanflight versions, access motors and switch to 4Way interface), the it will need only minor change.

I would expect that `pwmDisableMotors` will stop all access to ESC pins. We can improve this (implement pwmRelease / pwmReconfigure or similar)

**Comment by 4712** [comment]
@ledvinap MSP is used only as a frame, but not for any flash procedure.
Disabling the motor pwm is not the problem, but there should be as less as possible other ongoing actions, which could disturb (MSP packages ...).

ESC flashing and parameter handling will never be done in flight time.
So, switch off all disturbing procedure (as MSP, PWM and Beeper is in this case), do the job with the 4way interface - and reinitialize all necessary after done work.

**Comment by ledvinap** [comment]
@4712 : MSP packets should not pose any problem - BLHeliSuite does own serial port (no injected 'foreign' commands), there will be no change from BLHeli framing.
Other functions will be called, but only between commands, so it should not interfere with ESC programming. Interrupt handlers are called in both cases.
Beeper will be handled as usual - beep time may be prolonged by single CMD processing time (only chip erase is long enough to cause audible difference)

There is no acute need for  implementing this, but it would be nice in long-term.

**Comment by ledvinap** [comment]
@hydra, @4712 : this is ready for merge ...

**Comment by hydra** [comment]
ok, i will review and merge, maybe tonight if i get time.  I've just arrived in ibiza and have setup the office, I'll be here for 3 weeks and can now continue with reviewing PR's and merging and finishing up CF 1.13.0.  I am feeling the need to release CF 1.13.0 ASAP and move the IO EXTI 2 PR for the first thing for CF 1.14.0.

File: src/main/io/serial_4way.c [hunk 1]
```diff
@@ -333,511 +257,382 @@ uint16_t _crc_xmodem_update (uint16_t crc, uint8_t data) {
...
-        (pDeviceInfo->words[0] == 0xF410) || (pDeviceInfo->words[0] == 0xF390) || \
-        (pDeviceInfo->words[0] == 0xF850) || (pDeviceInfo->words[0] == 0xE8B1) || \
-        (pDeviceInfo->words[0] == 0xE8B2))
+static bool signatureMatch(uint16_t signature, uint16_t *list)
+{
+    for(; *list; list++)
+        if(signature == *list)
+            return true;
+    return false;
+}
```
Comment (hydra) [inline src/main/io/serial_4way.c]: yes much better.  wasn't a fan of the old device matching code at all.

**Comment by martinbudden** [comment]
@hyrda, I'm all for a v1.13.0 release. There are some bug related and other PRs that should make it into this release. For you convenience I've collected them together in issue https://github.com/cleanflight/cleanflight/issues/2103.

**Comment by hydra** [comment]
@martinbudden great thanks I'll take a look.

**Comment by hydra** [comment]
@ledvinap thanks for the great cleanup!  I gave this a test on a couple of boards tonight and didn't encounter any issues, merged.

**Comment by 4712** [comment]
@ledvinap thanks a lot. I merged it into 1.12 and noticed you saved up some bytes (~900 bytes for the Naze board). Quite impressive!

**Comment by hydra** [comment]
@4712 @ledvinap teamwork = win! :smile: :+1:

**Pull request merged**